pkg/cmd/roachtest/mixed_version_jobs.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"time"
+)
+
+type backgroundFn func(ctx context.Context, u *versionUpgradeTest) error
+
+// A backgroundStepper is a tool to run long-lived commands while a cluster is
+// going through a sequence of version upgrade operations.
+// It exposes a `launch` step that launches the method carrying out long-running
+// work (in the background) and a `stop` step collecting any errors.
+type backgroundStepper struct {
+	// This is the operation that will be launched in the background. When the
+	// context gets cancelled, it should shut down and return without an error.
+	// The way to typically get this is:
+	//
+	//  err := doSomething(ctx)
+	//  ctx.Err() != nil {
+	//    return nil
+	//  }
+	//  return err
+	run backgroundFn
+
+	// Internal.
+	m      *monitor
+	cancel func()
+}
+
+func makeBackgroundStepper(run backgroundFn) backgroundStepper {
+	return backgroundStepper{run: run}
+}
+
+// launch spawns the function the background step was initialized with.
+func (s *backgroundStepper) launch(ctx context.Context, _ *test, u *versionUpgradeTest) {
+	s.m = newMonitor(ctx, u.c)
+	ctx, s.m.cancel = context.WithCancel(ctx)
+	s.m.Go(func(ctx context.Context) error {
+		return s.run(ctx, u)
+	})
+}
+
+func (s *backgroundStepper) stop(_ context.Context, t *test, _ *versionUpgradeTest) {
+	s.m.cancel()
+	if err := s.m.WaitE(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func backgroundTPCCWorkload(warehouses int, tpccDB string) backgroundStepper {
+	return makeBackgroundStepper(func(ctx context.Context, u *versionUpgradeTest) error {
+		cmd := []string{
+			"./workload fixtures import tpcc",
+			fmt.Sprintf("--warehouses=%d", warehouses),
+			fmt.Sprintf("--db=%s", tpccDB),
```

Comment:
How does this invocation know where to connect? This can't possibly work, right? You're not running CRDB on last node. Probably getting lucky with `--local`?

Also, if this node restarts, aren't you screwed?

Maybe the "right" thing to do here is to have a step that starts the job (async, using a stmt_timeout if possible) and then just check the job periodically? Looks like then you could do without the intricacies of the backgroundStepper in the first place.